### ISSUE TITLE
SqlTransaction and TransactionScope leak isolation level

### ISSUE THREAD

**Issue by madelson** [dotnet/SqlClient#96]
The TransactionScope class seems to "leak" it's isolation level to future queries on the same (pooled) connection.

I would expect the isolation level of the connection to be restored when the transaction ends or is disposed.

Here is a snippet which reproduces the behavior:

```c#
SqlConnection.ClearAllPools();
var conn = new SqlConnection(new SqlConnectionStringBuilder { DataSource = @".\sqlexpress", IntegratedSecurity = true }.ConnectionString);

Action printIsolationLevel = () =>
{
	var cmd = conn.CreateCommand();
	cmd.CommandText = @"SELECT CASE transaction_isolation_level 
						WHEN 0 THEN 'Unspecified' 
						WHEN 1 THEN 'ReadUncommitted' 
						WHEN 2 THEN 'ReadCommitted' 
						WHEN 3 THEN 'Repeatable' 
						WHEN 4 THEN 'Serializable' 
						WHEN 5 THEN 'Snapshot' END AS TRANSACTION_ISOLATION_LEVEL 
						FROM sys.dm_exec_sessions 
						where session_id = @@SPID";
	Console.WriteLine(cmd.ExecuteScalar());
};

conn.Open();
printIsolationLevel(); // "ReadCommitted"
conn.Close();

using (var scope = new TransactionScope(
					TransactionScopeOption.RequiresNew,
					new TransactionOptions { IsolationLevel = System.Transactions.IsolationLevel.Serializable }))
{
	conn.Open();
	printIsolationLevel(); // "Serializable"
	conn.Close();
}

conn.Open();
printIsolationLevel(); // "Serializable" !!?
conn.Close();
```

**Comment by divega** [comment]
As recently [announced in the .NET Blog](https://devblogs.microsoft.com/dotnet/introducing-the-new-microsoftdatasqlclient/), focus on new SqlClient features an improvements is moving to the new [Microsoft.Data.SqlClient package](https://www.nuget.org/packages/Microsoft.Data.SqlClient/). For this reason, we are moving this issue to the new repo at https://github.com/dotnet/SqlClient. We will still use https://github.com/dotnet/corefx to track issues on other providers like System.Data.Odbc and System.Data.OleDB, and general ADO.NET and .NET data access issues.

**Comment by madelson** [comment]
Thanks @divega . Hopefully this new package will provide a chance to fix wonky behaviors like this one!

**Comment by szalapski** [comment]
I am encountering this now too, and I am surprised it is still unresolved.  Any update?

**Comment by cheenamalhotra** [comment]
Additional repro: https://github.com/dotnet/SqlClient/issues/1098

**Comment by justinbhopper** [comment]
> in the snapshot pool it doesn't hurt anything.

Actually, it can hurt if you use that connection to run stored procedures that start their own transaction.  The stored procedure may not have used `SET TRANSACTION ISOLATION LEVEL`, expecting the default isolation level to be utilized.

I think the confusion of all of this comes from the fact that the isolation level can be set via the construction of a disposable object (the transaction), rather than just a call directly on the connection.

Confusing:
```
using (var transaction = connection.BeginTransaction(IsolationLevel.Snapshot))
{
  ...
} // By now the transaction and isolation level should be reverted
```

Not confusing:
```
using (var transaction = connection.BeginTransaction(IsolationLevel.Snapshot))
{
  connection.SetIsolationLevel(IsolationLevel.Snapshot);
  ...
} // Now it is more clear that the isolation level was not related to the transaction's lifecycle
```

**Comment by nycdotnet** [comment]
@justinbhopper agree that is also a problem, but that's a whole separate issue which would cause pain even if the isolation level did not stick to a disposed connection.

**Comment by tbolon** [comment]
Also related: #146

**Comment by rgroenewoudt** [comment]
We are still running into this problem, without using TransactionScope.

**Comment by szalapski** [comment]
Agreed, I think this should be fixed, at least with an option at connection instantiation time to reset the "new" (reused) connection to the database-default isolation level.  It isn't confined to TransactionScope.

**Comment by michaelmesser** [comment]
This seems like a major bug. It has been 8 years.  Can this be fixed?

I would guess that many people might be affected by this and not realize it.

**Comment by itsvse** [comment]
I have also encountered this issue and hope to fix it as soon as possible. Thank you very much.
